pkg/spanner/migration.go

```diff
@@ -251,3 +279,90 @@ func isPartitionedDMLOnly(statement string) bool {
 func isDMLAny(statement string) bool {
 	return dmlAnyRegex.Match([]byte(statement))
 }
+
+// parseMigrationDirectives extracts migration directives in the format
+// @wrench.{key}={value} from the migration preamble.
+func parseMigrationDirectives(migration string) (MigrationDirectives, error) {
+	const (
+		directiveKeyPrefix = "@wrench."
+		migrationKindKey   = "migrationKind"
+		concurrencyKey     = "concurrency"
+	)
+
+	var directives MigrationDirectives
+	for _, line := range extractPreamble(migration) {
+		if !strings.HasPrefix(line, directiveKeyPrefix) {
+			continue
+		}
+
+		key, value, found := strings.Cut(line, "=")
+		if !found {
+			return MigrationDirectives{}, fmt.Errorf("directive must be in format @wrench.{key}={value}, got: %s", line)
+		}
+		value, _, _ = strings.Cut(value, " ")
+
+		switch strings.TrimPrefix(key, directiveKeyPrefix) {
+		case migrationKindKey:
+			migrationKind := MigrationKind(value)
+			if migrationKind != MigrationKindFixedPointIterationDML {
+				return MigrationDirectives{}, fmt.Errorf("invalid migration kind %q", migrationKind)
+			}
+			directives.MigrationKind = migrationKind
+		case concurrencyKey:
+			concurrency, err := strconv.Atoi(value)
+			if err != nil {
+				return MigrationDirectives{}, fmt.Errorf("invalid concurrency value %q", value)
+			}
+			directives.Concurrency = concurrency
+		default:
+			return MigrationDirectives{}, fmt.Errorf("unsupported directive: %s", key)
+		}
+	}
+	return directives, nil
+}
+
+// extractPreamble returns all comments from the start of a migration file,
+// until the first non-empty non-comment line is encountered.
+func extractPreamble(migration string) []string {
+	const (
+		blockCommentStart = "/*"
+		blockCommentEnd   = "*/"
+		lineCommentPrefix = "--"
```

Comment:
Do the block comments work in a migration? Wrench strips the line comments, can't remember if it's because spanner didn't support it at the time or maybe it only works in DML and not DDL api.

Reply:
Hmm, I didn't consider that perhaps one of the APIs does not support them.

The parser in general should.
They're definitely supported in the console, and more generally supported as part of GoogleSQL

https://cloud.google.com/spanner/docs/reference/standard-sql/lexical#multiline_comments

Reply:
Yeah ok, block comments seemingly aren't supported by the DDL API.

```
Failed to execute migration, Error parsing Spanner DDL statement: /*\n test block comment\n */\nCREATE TABLE TestTable (\n    Foo STRING(16) NOT NULL,\n) PRIMARY KEY(Foo) : Syntax error on line 1, column 1: Encountered \'/\' while parsing: ddl_statement
```

> Wrench strips the line comments, can't remember if it's because spanner didn't support it at the time or maybe it only works in DML and not DDL api

Hmm, when I try executing migrations via `gcloud spanner databases ddl update`, Spanner accepts line comments starting with `--`, but rejects line comments starting with `#` and rejects block comments.

I tested DML with block (and "#" line) comments in wrehcnh, but since they aren't stripped they're incorrectly identified as DDL by wrench (since they don't match wrench's DML regex).

I think it's probably best to just strip all valid comment types in wrench.

There's a method to we can lift from the `go-sql-spanner` pkg
https://github.com/googleapis/go-sql-spanner/blob/076c63111370017133f79dd37c0069f68f27d7df/statement_parser.go#L64-L68

Reply:
All comments are now stripped as of #162